Docsite extraction

solnic opened the pull request:
This is a bigger overhaul of the `Learn` section, which is reorganized with its sources extracted into individual repos.

Legacy docs will still live under `/%{version}/*` path, the new docs will live under `/learn/%{gem-slug}/%{version}` ie `/learn/sql/3.2`.

## TODO

- [x] tweak sorting of nav elements so that it uses the same order as specified in `projects.yml` + `data.position` for non-project nav resources
- [x] restore legacy version switcher so that people can still access `3.0`, `4.0` and `5.0` docs
- [x] fix broken internal links ie `/5.0/learn/core` => `learn/core/5.0` etc.
- [x] add `projects:symlink` rake task
- [x] extract docsite sources
  - [x] rom-core
  - [x] rom-repository
  - [x] rom-changeset
  - [x] rom-sql
  - [x] rom-http
  - [x] rom-factory
- [x] refactor sidebar nav to work with the extracted docsites

ianks commented on lib/rom/site/project.rb:
```diff
@@ -8,6 +8,14 @@ class Project < Middleman::Docsite::Project
+      end
+
+      def latest_version
+        versions.max_by { |version| version[:value].to_f }[:value]
```
we could use `Gem::Version` here

solnic replied:
@ianks oh thanks that's a great suggestion. I'll be moving this method to middleman-docsite so I'll use `Gem::Version` there.

solnic commented:
@ianks @timriley @flash-gordon 👋 I'm finally done with this transition. I'd appreciate if you could launch it locally and review it:

```
git clone https://github.com/rom-rb/rom-rb.org.git
cd rom-rb.org
bundle install
nvm use `cat .node-version`
yarn install
bundle exec rake projects:symlink
bundle exec middleman s
```

You will see that:

* `Learn` has now only one section provided by the site source code, which is `Introduction`, it is based on previous `Introduction` + `Getting Started`
* `Advanced` section is gone, I moved its adapter guide to `Guides` section and `Explicit setup` is now `Core / Framework setup` provided by `rom/core` docsite - this makes much more sense because it must be versioned
* I fixed formatting and probably most of the broken links. We can fix remaining ones whenever somebody find it.
* Legacy `3.0`, `4.0` and `5.0` docs are still available, the version switcher can be found *below* the navigation
* Existing project docs provided by their docsite sources are now versioned, just like in dry-rb.org project BUT there is **no version switcher** for individual projects. I will add it later once we need it (which means once there's a major upgrade of one of the gems).

solnic commented:
> The version switcher also breaks things if you use it in certain contexts. For example, if I navigate to /learn/introduction/installation/, and switch to Older versions: 4.0, I get a 404. I don't think that's a very nice experience... ideally we shouldn't be providing anything that can cause a user to jump to an error page.

@timriley I reckon to fix this properly we'd have to make the switcher really smart and I don't think it's worth the effort. This switcher is now a legacy switcher, we need a completely new one for the new docs because docs are per-gem now, there's no global version switcher idea anymore. What do you think if we just make legacy switcher redirect you to the main docs page? Docs changed between versions and detecting which page is available in which docs version is tricky, also what do you do when there's no corresponding page in another version of docs and you're trying to switch? That's why I think doing the simplest thing makes sense, especially that these docs will be legacy.

solnic commented:
> Firstly, it doesn't tell you what the current version you're using is. I think that's pretty important for understanding the context you might be switching away from, especially given browsers don't show full paths in their URL bars by default anymore. For example, I actually thought 5.0 was the newest version, because that's what was showing in the version switcher, but the newest version was actually 5.1! (which I only realised after clicking into my browser location bar to see the URL in full).

@timriley yeah that's tough because the version switcher you see is the legacy switcher. We need a new one going forward, I just didn't implement it as I don't know how to do it exactly. Again, legacy switcher (the one we have now) uses the idea of a global version, which was a HUGE mistake but it is what it is. We need to remove it eventually. I left it there because people need access to old docs obviously.

Maybe we should just deploy current website under a different subdomain and call it a day....people will still be able to find old docs. We can even tell google that URLs changed ie:

`http://rom-rb.org/4.0/learn/introduction/` => `http://old.rom-rb.org/4.0/learn/introduction/`

timriley commented:
I wonder if instead of the switcher, we just have a smaller set of links at the bottom of the nav? Ie. “4.0 docs, 3.0 docs” etc. and they would all just link to the root of that respective legacy version?

Seems like it would eliminate confusion and also avoid us having to spin up a whole separate legacy site?

solnic commented:
@timriley this is a perfect example of how convoluted my brain is - I love this idea 🙇

solnic commented:
@flash-gordon @timriley @ianks I've finally done it. The updated website using docsites from individual repos is live 😄

A gentle reminder that I extracted docsites only for the projects that were previously documented, so:

- rom-core
- rom-repository
- rom-changeset
- rom-sql
- rom-http

I also added docsite to `rom-factory` with a placeholder index saying "TODO: write docs" 🙄 Hopefuly somebody will write these docs one day.

Anyhow, this is a major improvement and I hope it's gonna work well for us 😄
